Feature/backend: implemented compiler options and language checks in the request handler

AndriiSerbeniuk opened the pull request:
Only for c++ though. Implementing other languages shouldn't raise problems.

AndriiSerbeniuk committed:
Added ability to specify files dir from outside

AndriiSerbeniuk committed:
Worked on code review issues:
- compilation handling is now done in a separate function;
- improved error handling in the filework module, other minor improvements;

AndriiSerbeniuk committed:
implemented lang info storage #8

AndriiSerbeniuk committed:
implemented flags parsing and validation code #8

AndriiSerbeniuk committed:
impl flags and lang check in submit handler #8

AndriiSerbeniuk committed:
Merge branch 'develop' of https://github.com/ITA-Dnipro/CompilerWebApp into feature/backend

virtustron reviewed (dismissed):
Great!

valentynfaychuk commented on webapp/src/languages/flags_validation_tests.rs:
```diff
@@ -0,0 +1,25 @@
+
+#[test]
+fn prohibited_cpp_flags() {
+    let cpp_info = cpp::construct();
```
I see that in both `valid_cpp_flags` and `prohibited_cpp_flags` the `cpp_info` is constructed the same way and the user has no impact on what flags will be used, is this `cpp::construct()` function destined for tests only?

AndriiSerbeniuk replied:
cpp::construct() returns information about c++ language. This information is constant, but I'm returning it dynamically because I had problems making it actually ```static```, since it doesn't let to use constructors, and this means no advanced structures for flags storage.
In these tests I wrote flags that were supposed to be tested inside strings in validator calls, ```cpp_info``` is the source that validator uses for its work.

valentynfaychuk commented on webapp/src/languages/compiler_flag.rs:
```diff
@@ -0,0 +1,152 @@
+        // Don't hash by a values of keys
+        match self
+        {   
+            CompilerFlag::Undefined => 
```
What are the use cases for `Undefined` and `UndefinedFlag(value)`, why are they two separate types?

AndriiSerbeniuk replied:
```Undefined``` is supposed to represent any undefined flag, I plan to use it in ```LangInfo``` flags lists. ```static_info::cpp::construct()``` returns ```LangInfo``` that contains ```Undefined``` in blacklisted flags, which means that any undefined flag is prohibited.
```UndefinedFlag``` also contains values of this flag. To think of it, it's not necessary, but I added it to have a way to return undefined values back to the caller. I honestly find it a little excessive, and it caused issues along the way as I was writing it. Removing it would simplify validation, but I'll have to come up with another way to pass undefined flags' info, if it's even necessary.

valentynfaychuk commented on webapp/src/filework/mod.rs:
```diff
@@ -1,35 +1,70 @@
+    // If error ever happens on work with temp files it's not on the user, 
+    // so he should get "internal server error" here
+    let mut input_file_name = env::var("COMPILATION_TEMP_DIR")
+        .expect("COMPILATION_TEMP_DIR doesn't exist");
```
Do we really want the server to crash in the middle of its normal operation, is it possible to return a `Result<...>` from this function (and then respond with 500 or 501 if it returns `Err(...)`)?

AndriiSerbeniuk replied:
I wanted to ask what to do about situations like this.  Should we respond with internal server errors explicitly? Rocket already does this if a handler panics, but I suppose doing it manually doesn't strain the server as much.

valentynfaychuk commented on webapp/src/filework/mod.rs:
```diff
@@ -1,35 +1,70 @@
+{
+    match remove_file(filename)
+    {
+        Ok(_) => println!("[INFO]: Removed {:?}.", filename),
```
Can you create a macro `info!("removed {}", filename)`?
It can unwrap to something like `println!("[info] removed {} (filework/mod.rs:50)", filename)`

AndriiSerbeniuk replied:
Sure. Looks like it's a #9 issue, is it okay if it'll be worked on as part of #9? And left as is for this pull request.

valentynfaychuk commented on webapp/src/http_handlers/submit/mod.rs:
```diff
@@ -1,34 +1,66 @@
+    if langs_info.contains_key(&compilation_data.lang)
+    {
+        let lang_info = &langs_info[&compilation_data.lang];
+        let mut validator = FlagsValidator::new();
```
I don't see the reason why the FlagsValidator is visible to this function.
Is it possible to move validation inside `try_to_compile` or embed validation into the serialization (or conversion to internal representation) of flags?

AndriiSerbeniuk replied:
Moving it into `try_to_compile` woud be right, don't know why I didn't do that in the first place.
I see how embeding the entire validation into rocket's data guard could be done by implementing custom `FromRequest` and a `Catcher` for it. I will look into it, but I'm not sure I'd be able to finish it for the demo

AndriiSerbeniuk committed:
Worked on code review issues:
- files creation now can't panic in runtime;
- post_submit() now responds with InternalServerError explicitly;
- moved validator instantiation into try_to_compile().

AndriiSerbeniuk committed:
Merge branch 'develop' of https://github.com/ITA-Dnipro/CompilerWebApp into feature/backend

AndriiSerbeniuk commented:
I'll also update the backend for newer compiler commits

AndriiSerbeniuk committed:
Updated backend for newer compiler commits

AndriiSerbeniuk committed:
Removed -v from blacklisted c++ flags
